### PR TITLE
Allow users to exhaust all their budget

### PR DESCRIPTION
**Pull request by leonardoarcari**
## Pull request details

This PR adds a new option to the `advanced` allocation algorithm to try to use _all budget_ available, even if it overweighs some assets.

The algorithm is roughly described as follows:

- Allocate budget as of today
- If some budget is left and `use_all_budget == true`:
  - Try to spread budget over under-allocated assets
  - Then, try to spread budget over all assets where we already allocated some cash, weighted by their target allocation
  - Finally, just drop all cash left, prioritizing the asset we allocated most cash over 

### Example

**Test case** : [dcapal-portfolio-use_all_budget-low.json](https://github.com/leonardoarcari/dcapal/files/13486835/dcapal-portfolio-use_all_budget-low.json)
**Budget** : €250

#### Without `use_all_budget`

image

#### With `use_all_budget`

image

## Breaking changes

_None_

## Issues fixed

1. #155

### CODE REVIEW

File: dcapal-optimizer-wasm/src/optimize/advanced.rs [hunk 1]
```diff
@@ -282,21 +306,103 @@ impl Problem {
...
         debug!(
-            "[Solution] budget_left={} open_assets={:?} w_sum={} adjusted_weights={:?}",
+            "[End] budget_left={} open_assets={:?} w_sum={} adjusted_weights={:?}",
             budget_left, open_assets, w_sum, adjusted_weights
         );
+    }
 
-        // Reconcile solution weights
-        for asset in solution.assets.values_mut() {
+    fn exahust_budget_left(
```
Comment (mattiapitossi) [inline dcapal-optimizer-wasm/src/optimize/advanced.rs:314]: Typo: it should be exhaust

**Review by corradopetrelli — APPROVED** [review]

File: dcapal-optimizer-wasm/src/optimize/advanced.rs (hunk 1 above)
Comment (leonardoarcari) [reply dcapal-optimizer-wasm/src/optimize/advanced.rs]: Thanks! Fixed

File: dcapal-optimizer-wasm/src/optimize/advanced.rs [hunk 2]
```diff
@@ -282,21 +306,103 @@ impl Problem {
...
+            asset.shares += allocated_shares;
             asset.weight = (asset.amount / pfolio_amount).round_dp(PERCENTAGE_DECIMALS);
+            *budget_left -= allocated;
         }
+        debug!("[Exhaust/1] After: budget_left={budget_left} candidates={candidates:?}");
 
-        solution.is_solved = true;
-        solution.budget_left = budget_left;
+        // Then, spread remaining budget by target weight
+        let mut candidates = assets
```
Comment (mattiapitossi) [inline dcapal-optimizer-wasm/src/optimize/advanced.rs:358]: Sorry, maybe I'm missing something here. But if all the budget was exhausted in the first step, will the following code still be executed?

Comment (leonardoarcari) [reply dcapal-optimizer-wasm/src/optimize/advanced.rs:358]: Oh right, good point. I should probably check for insignificant budget left and early return in that case. Thanks!

**Pull request merged**